openedx/core/djangoapps/courseware_api/views.py

```diff
@@ -310,7 +312,10 @@ def get(self, request, usage_key_string, *args, **kwargs):
         """
         Return response to a GET request.
         """
-        usage_key = UsageKey.from_string(usage_key_string)
+        try:
+            usage_key = UsageKey.from_string(usage_key_string)
+        except InvalidKeyError:
+            raise NotFound("Invalid usage key: '{}'.".format(usage_key_string))
```

Comment:
This seems sane, but would it be possible to add some context to the PR around what was happening?  What was being returned to the MFE prior to this?  (http status code, error messages, etc.)

Is InvalidKeyError the only thing we want to handle in this way?

Reply:
@davidjoy I apologize for the delay. I will get you an answer as soon as I can.

Reply:
@davidjoy It turns out this patch is not _strictly_ necessary: with the MFE patch applied (which it has been), the exception does not occur because the "null" never gets sent by the MFE to the LMS. I had to locally back out the MFE fix before I could reproduce the exception. However,  I do recommend the PR as a hand-crafted url could crash the server due to the unexpected exception.

The InvalidKeyError exception was cased by the MFE requesting `http://localhost:2000/course/course-v1:blank+1234+2020/null` thus the key was literally null. I had done the work for this PR prior to working on the MFE, thus I had been unaware of just what the MFE was doing incorrectly.

It appears the lms server was returning 200 at all times right up to the exception itself.

I am fairly confident that InvalidKeyError is the only exception that needs to be handled as my understanding is that opaque keys are (more or less) free-form strings used to index dictionaries.

The last several log lines before the exception are:
```
edx.devstack-master.lms   | [29/Jul/2020 04:14:05] "GET /api/courses/v2/blocks/?course_id=course-v1%3Ablank%2B1234%2B2020&username=edx&depth=3&requested_fields=children%2Cshow_gated_sections%2Cgraded HTTP/1.1" 200 495
edx.devstack-master.lms   | 2020-07-29 04:14:05,955 WARNING 316 [openedx.core.djangoapps.content.course_overviews.models] [user 3] models.py:779 - Falling back on modulestore to get course information for course-v1:blank+1234+2020
edx.devstack-master.lms   | [29/Jul/2020 04:14:06] "GET /api/courseware/course/course-v1:blank+1234+2020 HTTP/1.1" 200 1834
edx.devstack-master.lms   | 2020-07-29 04:14:06,514 INFO 316 [tracking] [user 3] logger.py:49 - {"username": "edx", "page": null, "ip": "172.26.0.1", "host": "localhost:18000", "event_type": "/api/courseware/resume/course-v1:blank+1234+2020", "referer": "http://localhost:2000/course/course-v1:blank+1234+2020/", "event_source": "server", "accept_language": "en-US,en;q=0.5", "agent": "Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0", "context": {"user_id": 3, "org_id": "", "path": "/api/courseware/resume/course-v1:blank+1234+2020", "course_id": ""}, "event": "{\"GET\": {}, \"POST\": {}}", "time": "2020-07-29T04:14:06.514672+00:00"}
edx.devstack-master.lms   | 2020-07-29 04:14:06,516 INFO 316 [tracking] [user 3] logger.py:49 - {"username": "edx", "page": null, "ip": "172.26.0.1", "host": "localhost:18000", "event_type": "/courses/course-v1:blank+1234+2020/masquerade", "referer": "http://localhost:2000/course/course-v1:blank+1234+2020/", "event_source": "server", "accept_language": "en-US,en;q=0.5", "agent": "Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0", "context": {"user_id": 3, "course_user_tags": {}, "org_id": "blank", "path": "/courses/course-v1:blank+1234+2020/masquerade", "course_id": "course-v1:blank+1234+2020"}, "event": "{\"GET\": {}, \"POST\": {}}", "time": "2020-07-29T04:14:06.516822+00:00"}
edx.devstack-master.lms   | [29/Jul/2020 04:14:06] "GET /api/courseware/resume/course-v1:blank+1234+2020 HTTP/1.1" 200 50
edx.devstack-master.lms   | 2020-07-29 04:14:06,824 INFO 316 [tracking] [user 3] logger.py:49 - {"username": "edx", "page": null, "ip": "172.26.0.1", "host": "localhost:18000", "event_type": "/api/courseware/sequence/null", "referer": "http://localhost:2000/course/course-v1:blank+1234+2020/null", "event_source": "server", "accept_language": "en-US,en;q=0.5", "agent": "Mozilla/5.0 (X11; Linux x86_64; rv:68.0) Gecko/20100101 Firefox/68.0", "context": {"user_id": 3, "org_id": "", "path": "/api/courseware/sequence/null", "course_id": ""}, "event": "{\"GET\": {}, \"POST\": {}}", "time": "2020-07-29T04:14:06.824597+00:00"}
edx.devstack-master.lms   | 2020-07-29 04:14:06,887 ERROR 316 [root] [user None] signals.py:23 - Uncaught exception from None
edx.devstack-master.lms   | Traceback (most recent call last):
edx.devstack-master.lms   |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.5/site-packages/opaque_keys/__init__.py", line 194, in from_string
edx.devstack-master.lms   |     namespace, rest = cls._separate_namespace(serialized)
edx.devstack-master.lms   |   File "/edx/app/edxapp/venvs/edxapp/lib/python3.5/site-packages/opaque_keys/__init__.py", line 225, in _separate_namespace
edx.devstack-master.lms   |     raise InvalidKeyError(cls, serialized)
edx.devstack-master.lms   | opaque_keys.InvalidKeyError: <class 'opaque_keys.edx.keys.UsageKey'>: null
```

Reply:
@davidjoy A slight correction, it turns out that any invalid key will cause the exception, eg: http://localhost:2000/course/course-v1:blank+1234+2020/razzledazzle